ibis-server/app/mdl/rewriter.py

```diff
@@ -1,21 +1,50 @@
 import httpx
 import orjson
+import sqlglot
 
 from app.config import get_config
-from app.logger import log_rewritten
+from app.logger import get_logger
+from app.model.data_source import DataSource
 
 wren_engine_endpoint = get_config().wren_engine_endpoint
 
+logger = get_logger("app.mdl.rewriter")
 
-@log_rewritten
-def rewrite(manifest_str: str, sql: str) -> str:
-    try:
-        r = httpx.request(
-            method="GET",
-            url=f"{wren_engine_endpoint}/v2/mdl/dry-plan",
-            headers={"Content-Type": "application/json", "Accept": "application/json"},
-            content=orjson.dumps({"manifestStr": manifest_str, "sql": sql}),
-        )
-        return r.text if r.status_code == httpx.codes.OK else r.raise_for_status()
-    except httpx.ConnectError as e:
-        raise ConnectionError(f"Can not connect to Wren Engine: {e}")
+
+class Rewriter:
+    def __init__(self, manifest_str: str, data_source: DataSource = None):
+        self.manifest_str = manifest_str
+        self.data_source = data_source
+
+    def rewrite(self, sql: str) -> str:
+        try:
+            r = httpx.request(
+                method="GET",
+                url=f"{wren_engine_endpoint}/v2/mdl/dry-plan",
+                headers={
+                    "Content-Type": "application/json",
+                    "Accept": "application/json",
+                },
+                content=orjson.dumps({"manifestStr": self.manifest_str, "sql": sql}),
+            )
+            rewritten_sql = (
+                r.text if r.status_code == httpx.codes.OK else r.raise_for_status()
+            )
+            logger.debug(f"Rewritten SQL: {rewritten_sql}")
+            return (
+                rewritten_sql
+                if self.data_source is None
+                else self.transpile(rewritten_sql)
+            )
+        except httpx.ConnectError as e:
+            raise ConnectionError(f"Can not connect to Wren Engine: {e}")
+
+    def transpile(self, rewritten_sql: str) -> str:
+        transpiled_sql = sqlglot.transpile(
+            rewritten_sql, read="trino", write=self.data_source.name
+        )[0]
+        transpiled_sql = sqlglot.parse_one(
+            transpiled_sql, dialect=self.data_source.name
+        ).sql(self.data_source.name)
```

Comment:
Curiously, why we need this line? I do some tests, then I found the result of `sqlglot.parse_one().sql()` is equal to `sqlglot.transpile`? There're any additional check in this api?

Reply:
Weird, I found a problem if only using `sqlglot.transpile` will get a SQL with invalid syntax `'` like `'WITH ABC ....'`. But now I re-test again, and it is gone.
I will remove this code. Thanks.